clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java

```diff
@@ -130,8 +130,13 @@ private MemoryRecordsBuilder convertRecordBatch(byte magic, ByteBuffer buffer, R
 
         MemoryRecordsBuilder builder = MemoryRecords.builder(buffer, magic, batch.compressionType(),
                 timestampType, recordBatchAndRecords.baseOffset, logAppendTime);
-        for (Record record : recordBatchAndRecords.records)
-            builder.append(record);
+        for (Record record : recordBatchAndRecords.records) {
+            // Down-convert this record. Ignore the presence of headers when down-converting to V0 and V1 (see KAFKA-6739).
```

Comment:
This is a bit subjective, but I personally prefer to avoid references to tickets unless they have information that is hard to replicate. In this case, the main point is that `append` will reject a record with headers if the message format doesn't support them, but that it the desired behaviour in this case. So we can just explain it here and the test you added makes sure we don't regress.

Reply:
The reason for this is that you get the context without having to load a separate page _and_ it's less likely to become stale as the code moves around (comments get stale, but references to external resources are even worse.)

Reply:
I tend to agree. The JIRA number doesn't add extra information that the comment itself couldn't provide. I think it's clear enough if we simply mention that we do not include headers since older message format versions don't support them.